docs: Add a bunch of documentation on Travis CI.

timabbott opened the pull request:
@adnrs96 @gnprice I'd love a review on this for clarity!

lonerz commented on docs/travis.md:
```diff
@@ -0,0 +1,102 @@
+  * For the production suite, this is a more complicated process
+    because of all the packages Travis installs.  See the comments in
+    `tools/travis/setup-production` for details.
+* `tools/travis/$TEST_SUITE`: A zulip script containing the actual
```
r/zulip/Zulip

timabbott commented:
> it seems like you use two spaces between sentences often.

Yep, that's what I learned in school (it was the standard taught in America at the time).  It's possible I should evaluate investing time to change my habits on this front.

adnrs96 commented on docs/travis.md:
```diff
@@ -0,0 +1,102 @@
+  helpful when debugging something complicated.
+
+[travis-fork]: git-guide.html#step-3-configure-travis-ci-continuous-integration
+
```
Shouldn't we mention about using `ts` to time tests here. I would add something like this after first point about performance issues.
* Also one can always use `ts` to profile the tests and figure out what runs
  slower. Checkout [`da731c`](https://github.com/zulip/zulip/commit/da731c)
  this commit to see how such a profiling could be done.

timabbott replied:
Yes, though thinking about this now, maybe we should just clean that change up to not break the exit status and just always run the tests through `ts`?

timabbott replied:
Let's open an issue for this idea.

timabbott replied:
Opened #5264.

adnrs96 commented on tools/travis/setup-production:
```diff
@@ -1,9 +1,17 @@
 #!/bin/bash
+# In short, this provisions a Zulip development environment and then
+# builds a Zulip release tarball (the same way we build them for an
+# actual release).  The actual test job wwill then install that.
```
r/wwill/will

adnrs96 commented on docs/travis.md:
```diff
@@ -0,0 +1,102 @@
+* If a test is failing nondeterministically in Travis CI, we consider
+  that to be an urgent problem.
+* If the tests become a lot slower, that is also an urgent problem.
+* Everything we do in CI should be to the extent possible runnable in
```
Something feels odd here around possible runnable. Maybe put a comma after possible.

gnprice replied:
Also it took me reading to the end of the sentence and coming back to realize this was meant to describe an alternate environment contrasting with Travis, not to call for things to run fast in Travis by e.g. being unit tests.

Also I think "to the extent possible" can be left out in favor of mentioning the one exception (the Travis config itself) later.

Here's another version:
* Everything we do in CI should also have a way to run it quickly (under 1 minute, preferably under 3 seconds), in order to iterate fast in development. Except when working on the Travis CI configuration itself, a developer should never have to repeatedly wait 10 minutes for a full Travis run.

gnprice commented on .travis.yml:
```diff
@@ -22,6 +30,9 @@ env:
 language: python
+# We do all builds for both Python 2.7 and 3.4, with the exception of
+# static analysis, which is just run once (and checks with both Python
+# versions).
```
"all suites", perhaps? The three things we're doing for each Python version are called "suites" here. On the Travis side, I think an actual running of the tests (a specific suite on a specific version) is a "job", and the thing that consists of one job per entry in this matrix is a "build".

Also I'd s/checks with/checks against/. "Checks with" sounds to me like it's using that version to do the checking.

gnprice commented on .travis.yml:
```diff
@@ -40,6 +51,9 @@ matrix:
 script:
+  # We unset GEM_PATH here as a hack to work around Travis CI having
+  # broken all ruby on their guests.  It's possible they fixed their
+  # bug and we can remove this.
```
Is there a Travis issue or other link for what this bug is?

timabbott replied:
no, they recommended this in a support thread.  My thinking is we should just do a follow-up project of removing it and seeing if things pass in this PR.

gnprice commented on docs/testing.md:
```diff
@@ -8,7 +8,8 @@ important components are documented in depth in their own sections:
 - [Node](testing-with-node.html): unit tests for JS front end code
-- [Linters](linters.html)
+- [Linters](linters.html): Our parallel linter suite
+- [Travis CI details](travis.html): How these run in Travis CI
```
s/these run/all of these run/

gnprice commented on docs/index.rst:
```diff
@@ -107,9 +107,10 @@ Contents:
    testing-with-django
+   testing-with-node
    testing-with-casper
+   travis
```
Not specific to this PR but it makes me notice that it's building up: the left nav is getting really long. Particularly unfortunate when using the left nav to browse sections that appear there below the fold -- when you get to the page, the part of the nav that you were just using disappears and you have to scroll down to get back to it.

I think probably the solution is another layer of collapsing detailed sections when not inside them -- e.g., "Code Testing" would just appear as one item in the left nav unless you're inside "Code Testing". Might motivate a bit of a round of tidying up the overall organization to ensure sections are still findable when their individual titles no longer appear on the nav.

timabbott replied:
Yeah, open an issue for this idea?

gnprice commented on docs/travis.md:
```diff
@@ -0,0 +1,102 @@
+The overall goal of our Travis CI setup is to catch as many possible
+future bugs in Zulip as possible, minimizing delay and false positives
+that waste the time of developers around the Zulip community
+debugging.  There are a few implications of this overall goal:
```
I want something like a preposition to connect "debugging".

"that waste the time of developers around the Zulip community through debugging"?
"... with debugging"?

gnprice commented on docs/travis.md:
```diff
@@ -0,0 +1,102 @@
+## Configuration
+
+The main Travis configuration file is `.travis.yml`.  The specific
+test suites we have are listed in the `matrix` section in that file.
```
Link to the file?

gnprice commented on docs/travis.md:
```diff
@@ -0,0 +1,102 @@
+test suites we have are listed in the `matrix` section in that file.
+We've configured it to use a few helper scripts for each job:
+
+* `tools/travis/setup-$TEST_SUITE`: A Zulip script that sets up the
```
The `TEST_SUITE` variable needs explanation here if you're going to refer to it.

Which seems like a good idea -- probably good to explain it anyway.

gnprice commented on docs/travis.md:
```diff
@@ -0,0 +1,102 @@
+We've configured it to use a few helper scripts for each job:
+
+* `tools/travis/setup-$TEST_SUITE`: A Zulip script that sets up the
+  test environment (E.g. installing dependencies).
```
comma after "e.g."

gnprice commented on .travis.yml:
```diff
@@ -40,6 +51,9 @@ matrix:
+  # broken all ruby on their guests.  It's possible they fixed their
+  # bug and we can remove this.
   - unset GEM_PATH
   - ./tools/travis/$TEST_SUITE
```
Reading this file to see how TEST_SUITE is used, I think I want `script` to be next to (right after) `install` -- they're logically quite similar.

Probably up at the top where `install` is now, since these are kind of the main event, though maybe `install` could move down here after where we set up TEST_SUITE.

gnprice commented on docs/travis.md:
```diff
@@ -0,0 +1,102 @@
+The main purpose of the distinction between the two is that if the
+`setup-backend` job fails, Travis CI will report it as a build
+infrastructure failure, whereas if the `backend` job fails, it'll be
+reported as a test failure.
```
Oh interesting! That's a good feature and actually makes the distinction more important than I'd realized -- I know Travis uses it to hide setup spew, but wasn't sure it did anything else.

Is this documented anywhere? I'm not finding any discussion on the web of a Travis infra-failure status. Even the [API docs](https://docs.travis-ci.com/api) don't say a word about what possible "build state" values are, beyond having some examples use "created", "failed", and "passed".

This bit should perhaps be a comment in the config. Like maybe
```
# How `install` differs from `script`:
# * Failure is reported by Travis as an infrastructure failure, rather than a test failure.
# * Output is hidden by default, because it's presumed to be boring.
```

timabbott replied:
They don't document it, it's just something you notice about the emails.  I'm going to leave it here, since I think it's useful higher-level explanation, though it could also be reasonable to add it to the `.travis.yml`

gnprice replied:
Yeah, definitely agree it belongs here. My wording was ambiguous but I meant it could be good to add there too.

gnprice commented on docs/travis.md:
```diff
@@ -0,0 +1,102 @@
+infrastructure failure, whereas if the `backend` job fails, it'll be
+reported as a test failure.
+
+An important detail is that Travis CI hides most build phases other
```
s/hides/hides by default output from/

gnprice commented on docs/travis.md:
```diff
@@ -0,0 +1,102 @@
+## Useful debugging tips and tools
+
+* For performance issues,
+  [this travis stats tool](https://scribu.github.io/travis-stats/#zulip/zulip/master)
```
s/travis/Travis/

gnprice commented on docs/travis.md:
```diff
@@ -0,0 +1,102 @@
+
+* You can [sign up your personal repo for Travis CI][travis-fork] so
+  that every remote branch you push will be tested, which can be
+  helpful when debugging something complicated.
```
Should we encourage this more prominently? I feel like I've heard you urge a number of people to do so, at the sprints and on chat.zulip.org.

Seems pretty handy -- the one downside I see is if it leads people to never explore running things faster locally, so we'd want to pair it with reinforcing that message.

timabbott replied:
It's in the git guide, but that's usually read by newer folks.  We probably need a step linked at the end of the development environment setup guides.

timabbott replied:
Opened #5266.

gnprice commented on docs/travis.md:
```diff
@@ -0,0 +1,102 @@
+production and the Zulip development environment) to ensure that each
+is named by a hash of its dependencies, so Zulip should always be
+using the same version of dependencies it would have used had the
+cache not existed.  In practice, bugs are always be possible, so be
```
s/always be/always/

gnprice commented on docs/travis.md:
```diff
@@ -0,0 +1,102 @@
+(effectively, because the Zulip installer does).  This carries a huge
+performance cost in Travis CI, because (1) they don't keep their test
+systems up to date and (2) literally everything is installed in their
+build works (e.g. several copies of Postgres, Java, MySQL, etc.).
```
s/works/workers/

gnprice commented on docs/travis.md:
```diff
@@ -0,0 +1,102 @@
+
+In order to make Zulip's tests performance reasonably well, we
+uninstall (or mark with `apt-mark hold`) many of these dependencies
+that are irrelevant to Zulip in `tools/travis/setup-production`.
```
link to source

gnprice commented on docs/travis.md:
```diff
@@ -0,0 +1,102 @@
+
+A consequence of this caching is that test jobs for branches which
+modify `package.json`, `requirements/`, and other key dependencies
+will be significantly slower than normal.
```
I'd spell this out a bit more: "than normal, because they won't get to benefit from the cache."

gnprice commented on tools/travis/backend:
```diff
@@ -1,4 +1,8 @@
+# This script is very similar to tools/test-all (what one runs
+# locally).  Possibly they should be merged, though it's worth noting,
+# they are intentionally different (basically some slow stuff is not
+# worth running in `test-all`).
```
Is this a suggestion to possibly make them behave the same (despite that slow stuff), or just to deduplicate the code?

If the latter, perhaps

>  This script is very similar to tools/test-all (what one runs locally). The intended difference is that we run here some slow stuff that's not worth running in `test-all`. Possibly the code should be deduplicated.

gnprice commented on tools/travis/setup-production:
```diff
@@ -1,9 +1,17 @@
+# This script is more complicated than that, basically because Travis
+# CI installs a ton of crap in its build workers, and we need to
+# remove some and reconfigure others to make things run smoothly and
+# quickly.
```
You have more explanation in the doc -- link to that?

gnprice commented:
Just sent a bunch of comments! This will be useful -- most of all, I think, the comments in the scripts and config.

Zooming out a bit: what are the intended uses of this document? It seems like a mix of developer docs and user docs, for users of Travis-for-Zulip. It'd probably be worth concentrating the user docs together so that people can read them without going through the developer docs. E.g., the "sign up your personal repo" bit, and maybe a section like "Reading Travis CI output" that discusses the setup/run structure and the hiding of output.

timabbott commented:
I pushed an updated version, that also features attempts to remove some of the code in `.travis.yml` that I think we can delete.  

Regarding purpose, this is intended primarily for developers working on our Travis CI setup who might want to know how it works.  I agree we should do some sort of extraction of the stuff useful to folks who don't care how it works to the top, though.  

BTW @umairwaheed FYI, is there anything I'm missing?  I guess we should mention coveralls?

Four follow-up issues to open:
* Extract "Using Travis" section to top of Travis CI doc.
* Make our `ts` stuff run unconditionally in Travis.
* https://github.com/zulip/zulip/pull/5247#discussion_r120453159
* https://github.com/zulip/zulip/pull/5247/files/65d39fb8997579afc4556a5133c8ed36a1704350#diff-8316f2fc56945c1095b64113e0fb0096

timabbott commented:
OK, I think I've addressed all the comments either in the form of an issue or just changing the docs.  Merged, thanks everyone for the reviews!  I expect we'll want to do several of those in the near term.

umairwaheed commented:
>BTW @umairwaheed FYI, is there anything I'm missing? I guess we should mention coveralls?

Yep, and the hooks that we call for Coveralls and Zulipbot. The Coveralls hook is used because we are using parallel builds. When a sub-build finishes it lets Coveralls know that it has finished.
